clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp

```diff
@@ -522,6 +522,80 @@ static Expected<StringRef> convertSPIRVToIR(StringRef Filename,
   return *TempFileOrErr;
 }
 
+// Update sycl-post-link options based on target triple.
+static void updateCmdArgs(SmallVector<StringRef, 8> &CmdArgs,
+                          llvm::Triple Triple) {
+  // Get an argument in CmdArgs that contains Str. If there is no such
+  // argument, an empty argument is returned
+  auto getArg = [&](const StringRef &Str) {
+    for (auto Arg : CmdArgs)
+      if (Arg.contains(Str))
+        return Arg;
+    return StringRef("");
+  };
+  // Add a new argument Arg to CmdArgs if not present already.
+  auto addArg = [&](const StringRef &Arg) {
+    auto ArgPresent = (getArg(Arg) != "");
+    if (!ArgPresent)
+      CmdArgs.push_back(Arg);
+  };
+  // Replace an argument in CmdArgs that contains Str with NewArg. If no such
+  // argument is present, add the NewArg to CmdArgs.
+  auto replaceOrAddArg = [&](const StringRef &NewArg, const StringRef &Str) {
+    for (auto &Arg : CmdArgs)
+      if (Arg.contains(Str)) {
+        Arg = NewArg;
+        return;
+      }
+    CmdArgs.push_back(NewArg);
+  };
+  // Remove argument containing Str from CmdArgs.
+  auto removeArg = [&](const StringRef &Str) {
+    CmdArgs.erase(
+        std::remove_if(CmdArgs.begin(), CmdArgs.end(),
+                       [&](StringRef Arg) { return Arg.contains(Str); }),
+        CmdArgs.end());
+  };
+
+  // specialization constants processing.
```

Comment:
I wonder if we can limit the cases where we have to remove or replace args as much as possible. In the ideal world, the driver would only pass these options as ClangLinkerWrapper options (and not any tool options) and ClangLinkerWrapper would know how to create the tool options based on the ClangLinkerWrapper options. Having the driver set tool options but then replacing/removing them in ClangLinkerWrapper seems like something we should try to minimize. It looks like we are stopping the driver from passing some options when the new offload driver is on, but I wonder if we plan to reduce the things this function is doing even more in the future.

Reply:
Thanks @sarnex. This is a good observation. I did spend some cycles trying to think of a better way. Following are the factors behind my current design. I will take -emit-only-kernel-as-entry-points as an example.
This option to sycl-post-link is controlled by two factors:
1. The -fsycl option passed by the user. This information is only available in the driver.
2. Target triple of the device image that is being post-linked. This information is embedded  in the device image and is parsed by the linker wrapper just before sycl-post-link. Again, this information is available only inside the linker wrapper.

We already have many sycl-specific options being passed into the linker wrapper. I think the trade-off here is between number of sycl-specific options passed into the linker wrapper versus adding logic to update options inside the '--sycl-post-link-options' based on the current target. I have gone with adding logic to update options.

May be we can document this and try to revisit at a later time if agreeable?

Thanks
Sincerely

Reply:
Thanks for the explanation. Let's assume for discussion `clang-linker-wrapper` has a way to get the offload kind and the option value (so it knows the current image is SYCL and it knows `emit-only-kernel-as-entry-points` was passed by the user to `clang++`). 

So I guess the discussion is, should these options be first-class to `clang-linker-wrapper` (meaning we need to add a new `clang-linker-wrapper` option for all of the SYCL options that effect device linking) or should they be opaque to `clang-linker-wrapper` (meaning they are blindly passed as a string from the driver and clang-linker-wrapper will have to do string manipulation on those options). My intuition is the first-class option is cleaner, but I'd like to get option from others (@bader @AlexeySachkov etc). 

This discussion doesn't necessarily have to block this PR but I think we should make a design decision at some point.

Reply:
I do not have strong opinion on either option. Surely, +1 on the fact that passing new options is a cleaner solution. I will create a tracker for adding new linker wrapper options for such cases. Hopefully, it need not block this PR.

Thanks

Reply:
I agree with @sarnex. clang-linker-wrapper doesn't seem like the right tool to update sycl-post-link options based on LLVM target triple. sycl-post-link tools itself should be able do that. Why can't we move this logic there?

Reply:
That sounds good to me, but I still would like to minimize string manipulation and rather combine triple information with option value information to decide behavior.